voyager.go

```diff
@@ -0,0 +1,327 @@
+	return &migrator{
+		db,
+		lockID,
+		lager.NewLogger("migrations"),
+		source,
+		migrationsRunner,
+		adapter,
+		&sync.Mutex{},
+	}
+}
+
+type migrator struct {
+	db                 *sql.DB
+	lockID             int
+	logger             lager.Logger
+	source             Source
+	goMigrationsRunner runner.MigrationsRunner
+	adapter            SchemaAdapter
+	*sync.Mutex
+}
+
+//go:generate counterfeiter . SchemaAdapter
+type SchemaAdapter interface {
+	MigrateFromOldSchema(*sql.DB) (int, error)
+	MigrateToOldSchema(*sql.DB, int) error
+	OldSchemaLastVersion() int
+}
+
+func (m *migrator) SupportedVersion() (int, error) {
+	matches := []migration{}
+
+	assets := m.source.AssetNames()
+
+	var parser = NewParser(m.source)
+	for _, match := range assets {
+		if migration, err := parser.ParseMigrationFilename(match); err == nil {
+			matches = append(matches, migration)
+		}
+	}
+	sortMigrations(matches)
+	return matches[len(matches)-1].Version, nil
+}
+
+func (m *migrator) CurrentVersion() (int, error) {
+
+	var migrationHistoryExists bool
+	err := m.db.QueryRow("SELECT EXISTS (SELECT 1 FROM information_schema.tables where table_name='migrations_history')").Scan(&migrationHistoryExists)
+	if !migrationHistoryExists && m.adapter != nil {
+		return m.adapter.OldSchemaLastVersion(), nil
+	}
+
+	var currentVersion int
+	var direction string
+	var dirty bool
+	err = m.db.QueryRow("SELECT version, direction, dirty FROM migrations_history WHERE status!='failed' ORDER BY tstamp DESC LIMIT 1").Scan(&currentVersion, &direction, &dirty)
+	if err != nil {
+		if err == sql.ErrNoRows {
+			return 0, nil
+		}
+		return -1, err
+	}
+
+	if dirty {
+		return -1, errors.New("could not determine current migration version. Database is in a dirty state")
+	}
+
+	for direction == "down" {
+		err := m.db.QueryRow("SELECT version, direction FROM migrations_history WHERE status!='failed' AND version < $1 ORDER BY tstamp DESC LIMIT 1", currentVersion).Scan(&currentVersion, &direction)
+		if err != nil {
+			return -1, multierror.Append(errors.New("could not determine current migration version"), err)
+		}
+	}
+	return currentVersion, nil
+}
+
+func (m *migrator) Migrate(toVersion int) error {
+
+	acquired, err := m.acquireLock()
+	if err != nil {
+		return err
+	}
+
+	if acquired {
+		defer m.releaseLock()
+	}
+
+	var existingDBVersion int
+	if m.adapter != nil {
+		existingDBVersion, err = m.adapter.MigrateFromOldSchema(m.db)
+		if err != nil {
+			return err
+		}
+	}
+
+	_, err = m.db.Exec("CREATE TABLE IF NOT EXISTS migrations_history (version bigint, tstamp timestamp with time zone, direction varchar, status varchar, dirty boolean)")
+	if err != nil {
+		return err
+	}
+
+	if existingDBVersion > 0 {
+		var containsOldMigrationInfo bool
+		err = m.db.QueryRow("SELECT EXISTS (SELECT 1 FROM migrations_history where version=$1)", existingDBVersion).Scan(&containsOldMigrationInfo)
```

Comment:
Same comments here as line `72`. There could be errors from `m.db.QueryRow(...).Scan()` which are lost with the side effect of jumping into the control flow statement below for reasons other than it containing old migration info.